### ISSUE TITLE
UploadDirectory should consider the FtpRules in Mirror Mode on Server side

### ISSUE THREAD

**Issue by philippjenni** [robinrodricks/FluentFTP#561]
Motiviation
--------------

I want to mirror application files (WebApp) to a webserver over ftp which has application data on the server that must not be deleted on update.

Description
--------------
If you will mirror a local directory to the server with FtpRules to ignore Directories, ignored directories should not be deleted on the server.

**Comment by robinrodricks** [comment]
I don't understand this. Can you explain. Currently we have full support for Mirror/Update modes on download and upload directory. What are you trying to implement here?

**Comment by philippjenni** [comment]
Here an example as use case

I have an ASP.NET application which has an App_Data Folder. The application on server stores data in this App_Data folder.

I have built a new Release of this application, that have an empty App_Data Folder. Now when i mirror the local folder to the server, the App_Data folder would be deleted, because the local directory it is empty. Now i have create an FtpRule that excludes the App_Data folder to upload. Files in this folder are no more uploaded, but the folder on the server is deleted by the uploading process.

It is a must,that the app_data folder on the server is not changed

**Comment by robinrodricks** [comment]
I think just a simple option that says DeleteFilteredFiles would be enough. It would control if the filtered out files are ignored or deleted.

**Comment by philippjenni** [comment]
> I think just a simple option that says DeleteFilteredFiles would be enough. It would control if the filtered out files are ignored or deleted. So this way you can add a rule to filter out the App_Data folder and it will not be deleted. It will be ignored.

Yes that be an good option ... if the option is not set it should be the current behavoir ... i can implement this, but should i change UploadDirectory's with a new parameter. I can't find a parameter where it makes sense to expand the option.

**Comment by philippjenni** [comment]
![Synchronisation Tasks](https://user-images.githubusercontent.com/579302/79630482-f589bc00-8151-11ea-8cc5-2a242a1ed797.png)

OK. But this should implement all the rules in the graph:

- Excluded files and folders should not be uploaded (App_Data, LogData.mdf)
- Excluded files and folders should not be deleted on remote side (LogData.ldf)
- Additional files and folders on local side should be uploaded (ImageMedium.png)
- Additional files and folders on remote side should be deleted (ImageSmall.png, file1.tmp, file2.tmp)
- Changed files should be overwritten (All yellow marked files)

This is what my pull request implemented

**Comment by robinrodricks** [comment]
Thanks for the nice diagram. I can implement #1 but #5 will be as per current rules - it will recopy only if file size mismatches. If you want to recopy if date mismatches it will have to be implemented.

**Comment by philippjenni** [comment]
5 is currently implemented. Thats OK

**Comment by robinrodricks** [comment]
#### 32.3.4
  - New: `UploadDirectoryDeleteExcluded` property to control if excluded files are deleted during Upload
  - New: `DownloadDirectoryDeleteExcluded` property to control if excluded files are deleted during Download
  - Fix: Dispose AsyncWaitHandles to stop handle leak in .NET Framework 4.5

Test if it works?

[FluentFTP.32.3.4.zip](https://github.com/robinrodricks/FluentFTP/files/4516351/FluentFTP.32.3.4.zip)

**Comment by robinrodricks** [comment]
Please inform me if it works?

**Comment by philippjenni** [comment]
Sorry, it doesn't work correctly ... Files and Folders there are excluded are removed. It looks that the problem is the path. I use this code to copy

```
string source = "C:\\Daten\\data";
string destination = "/subdomains/dev/httpdocs/ligamgr_ssha_prod";

client.UploadDirectory(source, destination, FtpFolderSyncMode.Mirror, FtpRemoteExists.Overwrite, FtpVerify.None, this.GetRulesFromFilter(source, destination, filter), itm => this.ReportProgress(itm))
```

The Exclude Rules works correctly, the files are not copied to the server. But extra files at the server in excluded folders are deleted.

**Comment by philippjenni** [comment]
I have analyzed your code. You check only the files to upload and not the files which only exists on the server. Files they are only exists on the server comes never to the "shouldExist" list.

**Comment by robinrodricks** [comment]
Thanks for checking. I have taken bits of your code and implemented it now. Try now?

[FluentFTP.32.3.4_Beta2.zip](https://github.com/robinrodricks/FluentFTP/files/4527619/FluentFTP.32.3.4_Beta2.zip)

**Comment by philippjenni** [comment]
Now it works. Thanks ... 2 points could still be improved

- UploadDirectoryDeleteExcluded and DownloadDirectoryDeleteExcluded are not part of the interface "IFtpClient"
- Empty folders on the target site are not deleted in mirror mode (I think this should be an extra ticket)

**Comment by robinrodricks** [comment]
I will fix the interface and release. Thanks again for your help.

**Comment by robinrodricks** [comment]
Released in [32.4.0](https://www.nuget.org/packages/FluentFTP/32.4.0)

#### 32.4.0
  - New: `UploadDirectoryDeleteExcluded` property to control if excluded files are deleted during Upload (thanks [philippjenni](/philippjenni))
  - New: `DownloadDirectoryDeleteExcluded` property to control if excluded files are deleted during Download (thanks [philippjenni](/philippjenni))
  - Fix: Dispose AsyncWaitHandles to stop handle leak in .NET Framework 4.5 (thanks [sdiaman1](/sdiaman1))
  - Fix: Implement proper cancellation support in `UploadDirectory` (once file transfer begins it cannot be cancelled)
  - Fix: Implement proper cancellation support in `DownloadDirectory` (once file transfer begins it cannot be cancelled)
  - Fix: Implement proper cancellation support in FXP `TransferDirectory`
  - Fix: Implement proper cancellation support in recursive `GetListing`
  - Fix: Correctly resume when unexpectedEOF error received during uploading a file (thanks [mrcopperbeard](/mrcopperbeard))
  - Fix: Hide internal properties in `FtpClient` that are not meant to be exposed
  - Fix: Update `IFtpClient` with the latest set of public properties that are meant to be exposed
